docs/smart-contracts/best-practices.md

```diff
@@ -0,0 +1,276 @@
+---
+title: Common mistakes and best practices 
+description: "Recommendations and reminders for patterns to avoid and practices to follow when writing programs using the Pact smart contract programming language." 
+id: best-practices
+---
+
+# Common mistakes and best practices 
+
+As a programming language, Pact has some unique features that make it adaptable to writing sophisticated applications, but that are also flexible enough to coding mistakes that can make a contract vulnerable to potential misuse. 
+This topic summarizes common mistakes to avoid and best practices you should follow as you develop programs with Pact. 
+
+## Common issues
+
+The following list summarizes the most common difficulties that developers who are new to Pact encounter or are most likely to misinterpret.
+
+- Internal methods are often not guarded by capabilities, allowing any unprivileged user to modify contract state.
+
+- Capabilities are often granted in cross-module calls by accident, giving the target module more permissions than it should have in the calling module.
+
+- How namespaces, accounts, and guards are used in Pact can be difficult to adjust to for developers familiar with other blockchain ecosystems or programming languages.
+  For example, Solidity developers are used to accounts being public keys and contracts having addresses, and often write smart contracts that only support single key (k:) accounts or deploy contracts insecurely in the free namespace.
+
+- Developers often forget to add types to their contracts, causing crashes in transactions later on.
+
+- Projects often split contract functionality into too many separate contracts—even if they share the same governance—complicating security and the overall design.
+
+## Best practices 
+
+The following list summarizes patterns, practices, and strategies for writing Pact code and delivering quality projects for the Kadena ecosystem.
+
+- Explicitly type all variables in all functions. 
+
+- Run the type checker over the code as you iterate on the implementation. 
+
+- Create unit tests for every function that fully exercise error cases. 
+
+- Be careful when using module references. 
```

Comment:
This needs far more elaboration, so I think this should have a link to some common pitfalls. This is a big one, and we should maybe emphasize that, because this has been the cause of several vulns in the past.

Reply:
We never covered specific examples (as far as I could find) in our brainstorming. The only specific one I'm aware of is the one you found in Marmalade.
The only additional info I found in the old doc was something like "Don't assume the module you reference can be trusted."
We don't have a list of common pitfalls to link to at this point.